### PR TITLE
Include reCAPTCHA for user validation

### PR DESCRIPTION
**Pull request by Leowisd**
Merge Recaptcha branch into Master, implemented https://github.com/oncokb/oncokb/issues/2099

Have fixed the footer clicking issue https://github.com/oncokb/oncokb-public/pull/579

### CODE REVIEW

File: src/main/java/org/mskcc/cbio/oncokb/security/uuid/UUIDFilter.java [hunk 1]
```diff
@@ -20,6 +20,7 @@
 public class UUIDFilter extends GenericFilterBean {
 
     public static final String AUTHORIZATION_HEADER = "Authorization";
+    public static final String RECAPTCHA_HEADER = "reCAPTCHA";
```
Comment (zhx828) [inline src/main/java/org/mskcc/cbio/oncokb/security/uuid/UUIDFilter.java:23]: We don't need this anymore.

File: pom.xml [hunk 2]
```diff
@@ -290,6 +291,11 @@
             <version>${archunit-junit5.version}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.github.mkopylec</groupId>
+            <artifactId>recaptcha-spring-boot-starter</artifactId>
+            <version>${recaptcha-spring-boot-starter.version}</version>
+        </dependency>
```
Comment (zhx828) [inline pom.xml:298]: We don't need this anymore.

File: pom.xml [hunk 3]
```diff
@@ -90,6 +90,7 @@
         <junit.utReportFolder>${project.testresult.directory}/test</junit.utReportFolder>
         <junit.itReportFolder>${project.testresult.directory}/integrationTest</junit.itReportFolder>
         <jslack.version>3.2.5</jslack.version>
+        <recaptcha-spring-boot-starter.version>2.3.1</recaptcha-spring-boot-starter.version>
```
Comment (zhx828) [inline pom.xml:93]: We don't need this anymore.

File: src/main/webapp/app/shared/auth/RecaptchaBoundaryRoute.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  isUserAuthenticated: boolean;
+  windowStore: WindowStore;
+}
+
+@observer
+export class RecaptchaBoundaryRoute extends React.Component<
+  IRecaptchaBoundaryRoute
+> {
+  @computed
+  get recaptchaVerifyed() {
```
Comment (zhx828) [inline src/main/webapp/app/shared/auth/RecaptchaBoundaryRoute.tsx:19]: Do we need this?

Comment (zhx828) [reply src/main/webapp/app/shared/auth/RecaptchaBoundaryRoute.tsx:19]: recaptchaVerified in windowStore is a observable already.

**Review by zhx828 — CHANGES_REQUESTED** [review]
Let's do a prettier on all files. I see some file changes which can be solved if we run the prettier.

**Comment by Leowisd** [comment]
> Let's do a prettier on all files. I see some file changes which can be solved if we run the prettier.

Do you mean we run prettier on master branch?

File: src/main/java/org/mskcc/cbio/oncokb/config/SecurityConfiguration.java [hunk 5]
```diff
@@ -29,6 +29,7 @@
 public class SecurityConfiguration extends WebSecurityConfigurerAdapter {
 
     private final TokenProvider tokenProvider;
+    // private final RecaptchaValidator recaptchaValidator;
```
Comment (zhx828) [inline src/main/java/org/mskcc/cbio/oncokb/config/SecurityConfiguration.java:32]: let's get rid of this.

File: src/main/webapp/app/App.tsx [hunk 6]
```diff
@@ -29,6 +33,15 @@ class App extends React.Component {
 
   constructor(props: IAppConfig) {
     super(props);
+    this.stores.windowStore.recaptchaRef = React.createRef();
+    this.stores.windowStore.recaptchaVerified = false;
```
Comment (zhx828) [inline src/main/webapp/app/App.tsx:37]: Shouldn't we create these in the windowStore instead?

File: src/main/webapp/app/App.tsx [hunk 7]
```diff
@@ -54,11 +67,29 @@ class App extends React.Component {
...
-        </Provider>
+        <>
+          <Reaptcha
+            ref={this.stores.windowStore.recaptchaRef}
+            sitekey={RECAPTCHA_SITE_KEY}
+            onVerify={this.onExecuteChange}
+            onRender={() => {
+              if (
+                !this.stores.authenticationStore.isUserAuthenticated &&
+                this.stores.routing.location.pathname !== '/'
```
Comment (zhx828) [inline src/main/webapp/app/App.tsx:78]: Let's use PAGE_ROUTE.HOME

File: src/main/webapp/app/indexUtils.tsx [hunk 8]
```diff
@@ -16,6 +17,14 @@ export const getStoredToken = () => {
   return Storage.local.get(AUTH_UER_TOKEN_KEY) || getPublicWebsiteToken();
 };
 
+export const setRecaptchaToken = (token: string) => {
+  return Storage.local.set(RECAPTCHA_KEY, token);
+};
+
+export const getRecaptchaToken = () => {
+  return Storage.local.get(RECAPTCHA_KEY);
+};
```
Comment (zhx828) [inline src/main/webapp/app/indexUtils.tsx:26]: Now if we only verify once, do we still need to save the recaptcha token?

File: src/main/webapp/app/shared/auth/RecaptchaBoundaryRoute.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  render() {
+    if (this.props.isUserAuthenticated) {
+      return <ErrorBoundaryRoute {...this.props} />;
+    } else {
+      if (!this.props.windowStore.recaptchaVerified) {
+        return (
+          <>
+            <div>
+              <h3 style={{ textAlign: 'center' }}>
+                Verifying your behavior...
```
Comment (zhx828) [inline src/main/webapp/app/shared/auth/RecaptchaBoundaryRoute.tsx:27]:
```suggestion
                Verifying your identity using Google reCAPTCHA...
```

File: src/main/webapp/app/store/WindowStore.ts [hunk 10]
```diff
@@ -18,6 +20,18 @@ class WindowStore {
     this.windowObj = window;
     this.setWindowSize();
     this.windowObj.addEventListener('resize', this.handleWindowResize);
+    this.windowObj.addEventListener('click', () => {
+      if (!this.recaptchaVerified && this.windowObj.location.pathname !== '/') {
```
Comment (zhx828) [inline src/main/webapp/app/store/WindowStore.ts:24]: here too, change the pathname

**Review by zhx828 — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/index.tsx [hunk 11]
```diff
@@ -22,6 +22,7 @@ import {
   assignPublicToken,
   getStoredToken,
   getPublicWebsiteToken,
+  AUTH_UER_TOKEN_KEY,
```
Comment (zhx828) [inline src/main/webapp/app/index.tsx:25]: This is no longer needed

File: src/main/webapp/app/store/WindowStore.ts [hunk 12]
```diff
@@ -10,6 +11,8 @@ export interface IWindowSize {
 
 class WindowStore {
   @observable size: IWindowSize;
+  public recaptchaRef: any;
+  @observable recaptchaVerified: boolean;
```
Comment (zhx828) [inline src/main/webapp/app/store/WindowStore.ts:15]: Can we put the observables together to keep the file structured?

**Review by zhx828 — CHANGES_REQUESTED** [review]
Looks good. Just two minor comments.

**Comment by zhx828** [comment]
We need to squash the commits before merging.

**Comment by Leowisd** [comment]
Made some error on squash commits on this branch, just reopen a new PR from Leowisd/recaptcha branch https://github.com/oncokb/oncokb-public/pull/591

**Pull request closed without merging**